### PR TITLE
feat: workflow as code, step 1

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/workflow/repository.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+				return nil
+			}
+			continue
+		}
+	}
+}
+
+func createOperationRequest(w sdk.Workflow, opts sdk.WorkflowRunPostHandlerOption) (sdk.Operation, error) {
+	ope := sdk.Operation{}
+	if w.Root.Context.Application == nil {
```
Comment (bnjjj) [inline engine/api/workflow/repository.go]: just to check, but here the context could be nil

Comment (sguiheux) [reply engine/api/workflow/repository.go]: Context cannot be nil after loading workflow from DB, each node must have a context

File: cli/cdsctl/workflow_as_code.go [hunk 2]
```diff
@@ -396,15 +401,15 @@ func workflowInitRun(c cli.Values) error {
 	}
 
 	if err := workflowTarReaderToFiles(dotCDS, tr, true, true); err != nil {
-		return err
+		return fmt.Errorf("error on workflowTarReaderToFiles:%s", err)
```
Comment (fsamin) [inline cli/cdsctl/workflow_as_code.go]: this message is not user-friendly, you should fix workflowTarReaderToFiles to raise a nice error
you should use %v to print an error (the `default format` as explained in https://golang.org/pkg/fmt/)

File: docs/content/gettingstarted/firstworkflow.md [hunk 3]
```diff
@@ -1,40 +1,47 @@
 +++
 title = "First Workflow from repository"
 weight = 1
-aliases = [
-    "/gettingstarted"
-]
 +++
```
Comment (fsamin) [inline docs/content/gettingstarted/firstworkflow.md]: why removing the aliases ? it was nice to not have an empty page au the root level

File: cli/cdsctl/workflow_as_code.go [hunk 4]
```diff
@@ -395,15 +401,15 @@ func workflowInitRun(c cli.Values) error {
...
 	}
 
 	//Configure local git
 	if err := gitRepo.LocalConfigSet("cds", "workflow", name); err != nil {
-		return err
+		return fmt.Errorf("error on LocalConfigSet:%s", err)
```
Comment (fsamin) [inline cli/cdsctl/workflow_as_code.go]: i prefer the standard git error returned by the lib. Naming the function is not very user friendly

File: cli/cdsctl/workflow_as_code.go [hunk 5]
```diff
@@ -395,15 +401,15 @@ func workflowInitRun(c cli.Values) error {
...
 	}
 	if err := gitRepo.LocalConfigSet("cds", "application", appName); err != nil {
-		return err
+		return fmt.Errorf("error on LocalConfigSet:%s", err)
```
Comment (fsamin) [inline cli/cdsctl/workflow_as_code.go]: i prefer the standard git error returned by the lib. Naming the function is not very user friendly

File: cli/cdsctl/workflow_as_code.go [hunk 6]
```diff
@@ -22,7 +22,11 @@ var workflowInitCmd = cli.Command{
 	Name:  "init",
 	Short: "Init a workflow",
 	Long: `[WARNING] THIS IS AN EXPERIMENTAL FEATURE
-Initialize a workflow from your current repository, this will create yml files and push them to CDS.`,
+Initialize a workflow from your current repository, this will create yml files and push them to CDS.
+
+Documentation: https://ovh.github.com/cds/gettingstarted/firstworkflow/
```
Comment (fsamin) [inline cli/cdsctl/workflow_as_code.go]: github.io

File: cli/cdsctl/workflow_as_code.go [hunk 7]
```diff
@@ -177,98 +181,99 @@ func workflowInitRun(c cli.Values) error {
...
+	}
 
-		// Try to find application or create a new one from the repo
-		apps, err := client.ApplicationList(pkey)
-		if err != nil {
-			return fmt.Errorf("unable to list applications: %v", err)
+	// Check it the project with it's delegated oauth knows the current repo
+	// Search the repo
+	var repoFound bool
+	fullname := strings.ToLower(fmt.Sprintf("%s/%s", fullnames[0], fullnames[1]))
```
Comment (fsamin) [inline cli/cdsctl/workflow_as_code.go]: repoName ?

File: cli/cdsctl/workflow_as_code.go [hunk 8]
```diff
@@ -177,98 +181,99 @@ func workflowInitRun(c cli.Values) error {
...
+	for _, r := range repos {
+		// r.Fullname = CDS/demo
+		if strings.ToLower(r.Fullname) == fullname {
```
Comment (fsamin) [inline cli/cdsctl/workflow_as_code.go]: fullname => repoName

File: cli/cdsctl/workflow_as_code.go [hunk 9]
```diff
@@ -177,98 +181,99 @@ func workflowInitRun(c cli.Values) error {
...
+				break
 			}
 		}
 	}
 
 	var pipName string
 	if shouldCreatePipeline {
-		fmt.Print("Enter your pipeline name : ")
+		fmt.Print("Enter your pipeline name: ")
 		pipName = cli.ReadLine()
```
Comment (fsamin) [inline cli/cdsctl/workflow_as_code.go]: we should check the name pattern

File: cli/cdsctl/workflow_as_code.go [hunk 10]
```diff
@@ -309,26 +309,32 @@ func workflowInitRun(c cli.Values) error {
...
-				"app-ssh-" + repoManagerName: exportentities.KeyValue{
-					Type: sdk.KeyTypeSSH,
-				},
 				"app-pgp-" + repoManagerName: exportentities.KeyValue{
 					Type: sdk.KeyTypePGP,
 				},
 			},
 		}
 
+		if !strings.HasPrefix(fetchURL, "https") {
```
Comment (fsamin) [inline cli/cdsctl/workflow_as_code.go]: check connectionType

File: docs/content/gettingstarted/firstworkflow.md (hunk 3 above)
Comment (yesnault) [reply docs/content/gettingstarted/firstworkflow.md]: the page is not empty anymore

File: cli/cdsctl/workflow_as_code.go (hunk 2 above)
Comment (yesnault) [reply cli/cdsctl/workflow_as_code.go]: fixed

File: cli/cdsctl/workflow_as_code.go (hunk 9 above)
Comment (yesnault) [reply cli/cdsctl/workflow_as_code.go]: fixed

**Pull request merged**